### DISCUSSION TITLE
TCP Source

### DISCUSSION THREAD

**Discussion by ssozonoff** [lf-edge/ekuiper#3108] [Q&A]
Hi.

Does anyone know if a connector exists for sourcing data directly from a local TCP port ?

Thanks,
Serge

**Comment by ngjaying** [comment]
There is no built-in TCP soruce now.

**Reply by ssozonoff** [reply]
OK thanks. I will try and write a GO plugin for this. I have been looking at the examples and reading but not clear what ekuiper expects for its internal data format and how that takes place. For example with JSON input at the source. If you have any pointers to some documentation I would be grateful.

**Reply by ngjaying** [reply]
This is a feature request https://github.com/lf-edge/ekuiper/issues/19. It would be great if you could make it a built-in source. We did some refactor in v2 source/sink implementation, the doc is not completed yet. You could take a look at internal/io/redis/redisSub.go which is also ingesting bytes. In the planner, source and format are separated, another operator (decode node) will take care of JSON decoding. So in the source, you just need to call ingest(ctx, messageBytes).
